### ISSUE TITLE
swig isn't generating target file to build PHP wrapper

### ISSUE THREAD

**Issue by mnguyenpdftron** [swig/swig#2113]
OS: Ubuntu 18.04/Linux
Target: PHP 7.2.10, PHP 8.0.13 
Problem: Current swig (master branch) doesn't generate PHP output file needed to build PDFTron's PHP wrapper (swig 4.0.2 and older versions did build successfully with PHP 7.2.10)
Steps to reproduce:
  - Extract the zip package and run the Shell script. The script in the zip package contains comments explaining the lower level steps
   $ sh ./Build_PDFTron_PHPWrapper.sh
[BuildScript.zip](https://github.com/swig/swig/files/7644205/BuildScript.zip)

**Comment by ojwb** [comment]
This is an intentional change - here's the CHANGES.current entry:

```
2021-05-04: olly
            [PHP] #1982 #1457 https://sourceforge.net/p/swig/bugs/1339/
            SWIG now only use PHP's C API to implement its wrappers, and no
            longer generates PHP code to define classes.  The wrappers should
            be almost entirely compatible with those generated before, but
            faster and without some previously hard-to-fix bugs.

            The main notable difference is SWIG no longer generates a .php
            wrapper at all by default (only if %pragma(php) code=... or
            %pragma(php) include=... are specified in the interface file).
            This also means you need to load the module via extension=...
            in php.ini, rather than letting the dl() in the generated
            .php wrapper load it (but dl() has only worked for command-line
            PHP for some years now).

            *** POTENTIAL INCOMPATIBILITY ***
```

**Comment by wsfulton** [comment]
@mnguyenpdftron can you please summarise what the problems are for those that don't download the script.

@ojwb Before the travis CI testing fell apart, we were testing PHP 7.0, 7.1, 7.2, 7.3, 7.4, 8.0. We've been ramping up the replacement testing on Github actions and is currently testing PHP 7.4 and 8.0 only atm. Does swig-4.1 have a target set of PHP versions that it will support and can we add those to Github Actions for testing? I'm not sure if that includes PHP 7.2, but presumably SWIG will aim to support whatever version of PHP is included in the LTS version of 18.04 (while this version of Ubuntu is in support).

**Comment by wsfulton** [comment]
@ojwb looks like our responses crossed. @mnguyenpdftron given the response from @ojwb and no generated php files should be expected, what is the incompatibility you see?

**Comment by ojwb** [comment]
Re supported versions, the manual currently says "Currently any PHP7 or PHP8 release should work", though obviously we can adjust that.

[PHP upstream currently support 7.3 (but only for a few more days), 7.4, 8.0 and 8.1](https://www.php.net/supported-versions.php) so 7.4 and up definitely ought be supported and tested. I've not actually tried 8.1 yet though as it's only a week old.

It probably makes sense to aim to support 7.2 and 7.3 too - while they won't have upstream support, they are present in stable distros where they presumably have some level of security support.

I think the significant structural change to wrapping using the C API probably argues for testing back to 7.0 like we used to have.  The changes for this were extensive so anything we can easily do to shake out problems seems beneficial, and we've previously seen problems that only manifested with a particular PHP release series but turned out not to be specific to that series (we just got "lucky" with other release series).

In terms of special code to support older releases, there's very little - looks like 3 macros/functions added in 7.3 or 7.4 which we provide versions of for older releases (plus some for PHP 7 vs PHP 8 C API changes).

**Comment by mnguyenpdftron** [comment]
> This is an intentional change - here's the CHANGES.current entry:
> 
> ```
> 2021-05-04: olly
>             [PHP] #1982 #1457 https://sourceforge.net/p/swig/bugs/1339/
>             SWIG now only use PHP's C API to implement its wrappers, and no
>             longer generates PHP code to define classes.  The wrappers should
>             be almost entirely compatible with those generated before, but
>             faster and without some previously hard-to-fix bugs.
> 
>             The main notable difference is SWIG no longer generates a .php
>             wrapper at all by default (only if %pragma(php) code=... or
>             %pragma(php) include=... are specified in the interface file).
>             This also means you need to load the module via extension=...
>             in php.ini, rather than letting the dl() in the generated
>             .php wrapper load it (but dl() has only worked for command-line
>             PHP for some years now).
> 
>             *** POTENTIAL INCOMPATIBILITY ***
> ```
@ojwb Thanks for your prompt reply.  Does it apply to PHP only? I didn't see it in Python, In fact, our Python wrapper is working fine.

**Comment by ojwb** [comment]
Yes (as the "[PHP]" tag indicates)

**Comment by wsfulton** [comment]
@ojwb, I've added in testing of 7.0-8.1 as suggested in commit 8432d3f89b09e6511370cfea1765fabf224e10d9. Looks okay, with just one Deprecated warning in the php_iterator testcase for 8.1.

**Comment by ojwb** [comment]
Thanks.

I had a quick look at the CI log for 8.1.  It's the code in the testcase which is triggering this, though anyone copying that code to make a wrapped C++ class usable as a PHP iterator would also hit this.

I think adding support for PHP return type hints would solve this, which is already #2027 (and I've already done some work on that).  We could probably add hints automatically for the types involved here.

**Comment by mnguyenpdftron** [comment]
@ojwb @wsfulton 
Here is the result after I updated build script to exclude php target file
PHP7 : Build was successful but got Segmentation fault (core dumped)  when running app
PHP8 : Build was NOT successful. 
 error: macro "ZVAL_STRINGL" passed 4 arguments, but takes just 3
     ZVAL_STRINGL(return_value, const_cast<char*>(temp$argnum.data()), temp$argnum.size(), 1);
error: ‘code’ was not declared in this scope
 # define SWIG_HANDLE_VALUE_ERROR_FOR_PHP8 code == SWIG_ValueError ? zend_ce_value_error :
/usr/include/php/20200930/Zend/zend_types.h:563:35: error: no match for ‘operator*’ (operand type is ‘zval {aka _zval_struct}’)
 #define Z_TYPE_P(zval_p)   Z_TYPE(*(zval_p)
[error.txt](https://github.com/swig/swig/files/7665150/error.txt)
)

**Comment by ojwb** [comment]
```
ZVAL_STRINGL(return_value, const_cast<char*>(temp$argnum.data()), temp$argnum.size(), 1);
```

In PHP5 `ZVAL_STRINGL` did take a 4th argument, but it no longer does.  There's no code like this in SWIG's PHP7+ typemaps so I suspect either:

* you're using `-php5` (or `-php` on an older SWIG where that means `-php5`) - use `-php7` instead if so
* or SWIG is finding the wrong typemaps for some reason - how to fix that depends why that's happening...
* or you have custom typemaps in your interface files which need updating for the PHP5 to PHP7 API changes

The third error looks like the same cause to me.

I've no idea about the second error - it doesn't make much sense to get an error about a non-declared variable in a line defining a macro.  I'd try and fix the other problem and see if this goes away, or becomes a clearer error.

**Comment by mnguyenpdftron** [comment]
@ojwb : you're right. I got ZVAL_STRINGL fixed now. I was using -php7 but custom typemaps didn't pickup correctly when I was experimenting with php8. It seems the only error now is 
 error: ‘code’ was not declared in this scope
` # define SWIG_HANDLE_VALUE_ERROR_FOR_PHP8 code == SWIG_ValueError ? zend_ce_value_error :`
[error2.txt](https://github.com/swig/swig/files/7666128/error2.txt)

**Comment by mnguyenpdftron** [comment]
error went away when I made this  change to the macro: https://github.com/swig/swig/blob/master/Lib/exception.i
`
#ifdef SWIGPHP
%{
#if PHP_MAJOR >= 8
#define SWIG_exception(code, msg) do { zend_throw_exception( \
    code == SWIG_TypeError ? zend_ce_type_error : \
    code == SWIG_ValueError ? zend_ce_value_error : \
    code == SWIG_DivisionByZero ? zend_ce_division_by_zero_error : \
    code == SWIG_SyntaxError ? zend_ce_parse_error : \
    code == SWIG_OverflowError ? zend_ce_arithmetic_error : \
    NULL, msg, code); SWIG_fail; } while (0)
#else
#define SWIG_exception(code, msg) do { zend_throw_exception( \
    code == SWIG_TypeError ? zend_ce_type_error : \
    code == SWIG_DivisionByZero ? zend_ce_division_by_zero_error : \
    code == SWIG_SyntaxError ? zend_ce_parse_error : \
    code == SWIG_OverflowError ? zend_ce_arithmetic_error : \
    NULL, msg, code); SWIG_fail; } while (0)
#endif
%}

#endif
`

**Comment by ojwb** [comment]
The existing code avoids repetition of all the cases that are in common so I'm reluctant to change it as you suggest when I can't see any reason why the existing code would fail like this.

The current code works fine for me locally and in the CI builds, so I think there must be something in your interface file triggering this somehow.  Can you upload the SWIG-generated _wrap.c or _wrap.cxx file somewhere?

**Comment by mnguyenpdftron** [comment]
@ojwb: attached are the files 

[php8wrap.zip](https://github.com/swig/swig/files/7671826/php8wrap.zip)

This is what I had after updating,  building and running our applications. (Note: we didn't have any issues with swig 4.0.2 and before)

PHP7
```
PHP Warning:  Declaration of Popup::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of FileAttachment::Create($arg1, $arg2, $arg3, $arg4) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of Ink::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of Circle::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of HighlightAnnot::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of LineAnnot::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of FreeText::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of Movie::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of RedactionAnnot::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of Screen::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of Sound::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of Square::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of Squiggly::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of StrikeOut::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of Text::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of Underline::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Warning:  Declaration of Watermark::Create($arg1, $arg2) should be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0

PDFNet is running in demo mode.
Permission: read
Segmentation fault (core dumped)
```

PHP8
```
PHP Fatal error:  Declaration of Popup::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of FileAttachment::Create($arg1, $arg2, $arg3, $arg4) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of Ink::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of Circle::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of HighlightAnnot::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of LineAnnot::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of FreeText::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of Movie::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of RedactionAnnot::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of Screen::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of Sound::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of Square::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of Squiggly::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of StrikeOut::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of Text::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of Underline::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0
PHP Fatal error:  Declaration of Watermark::Create($arg1, $arg2) must be compatible with Annot::Create($arg1, $arg2, $arg3) in Unknown on line 0

PDFNet is running in demo mode.
Permission: read
Segmentation fault (core dumped)
```

**Comment by ojwb** [comment]
Oh, I see the problem - the PHP8 version of the code here doesn't work, but that's hidden because it's incorrectly gated by the check `PHP_MAJOR >=8` but `PHP_MAJOR` is not a macro PHP's headers define and should be `PHP_MAJOR_VERSION`.  I'll fix the check and the PHP8 version of the code.

I guess you must be defining `PHP_MAJOR` yourself for some reason?

Can you come up with a reduced testcase for the segmentation fault?

**Comment by mnguyenpdftron** [comment]
Originally,  I only used  `PHP_MAJOR_VERSION`, but had to add `PHP_MAJOR ` to get it build.

**Comment by ojwb** [comment]
Ah, presumably you're using `-Wundef` or similar, which would give a warning for this:

```
warning: "PHP_MAJOR" is not defined, evaluates to 0 [-Wundef]
```

You can stop defining this now.
